### PR TITLE
feat: custom transform for bundling and running suites

### LINKED ISSUE #596: create separate bundles for external modules
+ With the support of bundling external packages, we bundle the included packages along with the source journey files to be able to run them via Heartbeat later without installing any packages. This is good, but it changes the source line numbers and column numbers of those particular journeys and might be a bad UX if we have to go back to the original source. 

More details in here - https://github.com/elastic/synthetics/pull/580#pullrequestreview-1082955330

The ultimate goal of our push command is to replicate exactly how your folder structure is on your local and run the same inside HB when pushed. This would make it presentable in Kibana when we add unzipping support and allow users to view and edit monitors.

TLDR: If i push below example journey code 
~~examples/bundled-code/test.journey.ts~~
```ts
import { journey, step, monitor, expect } from "@elastic/synthetics";
import isPositive from "is-positive";
import leftpad from "leftpad";
import * as utils from "./utils";

journey("example journey", ({ page, params }) => {
  step("example step", () => {});
});
```

Our bundled zip folder should be in the below format. 

```
examples/bundled-code/test.journey.ts
examples/bundled-code/utils.ts
examples/bundled-code/is-positive.js
examples/bundled-code/leftpad.js
```

### LINKED ISSUE #652: journey did not finish executing, 0 steps ran when calling exported function.
Hi there!

I'm trying to structure my synthetic tests project by extracting methods (API requests mostly) to dedicated TS files. The problem is that I often hit the issue where my tests run and pass locally (with `npm test`), but the version I push to ELK's Uptime app (our ELK cloud is running 8.4.3) does not run at all. The monitor appears in a `Failed` state, and when I go to the details, I see 

> There are no steps for this journey
This journey did not contain any steps.
The journey's check group is 3388a32c-642b-11ed-9db9-ce3a36e90cb7.
There is no further information to display.

Under the history section, in the error column, I can see the following:

```
journey did not finish executing, 0 steps ran: command '/tmp/elastic-synthetics-unzip-20303004/node_modules/.bin/elastic-synthetics /tmp/elastic-synthetics-unzip-20303004/node_modules/.bin/elastic-synthetics /tmp/elastic-synthetics-unzip-20303004 --playwright-options {"headless":true,"ignoreHTTPSErrors":false} --screenshots off --throttling {"download":5,"latency":20,"upload":3} --rich-events --match My Example Journey --params "{1 hidden params}"' exited unexpectedly with code: 1
```

I manage to create a primer with exactly what seems to cause this failure, using the [base template](https://github.com/elastic/synthetics/tree/main/templates), with the following files in the `journeys` directory:

```
example.journey.ts
foo.ts
get-foo.ts
```

With the following content:

```typescript
// foo.ts
export class Foo {
    private readonly bar: string;
}


// get-foo.ts
import { Foo } from './foo';

export function createFoo(): Foo {
    return new Foo();
}


// example.journey.ts
import { expect, journey, monitor, step } from '@elastic/synthetics';
import { createFoo } from './get-foo';

journey('Journey foo no params', ({ page, params }) => {
    // Only relevant for the push command to create
    // monitors in Kibana
    monitor.use({
        id: 'example-monitor',
        schedule: 10,
        screenshot: 'off',
    });

    step('getFoo()', async () => {
        const foo = createFoo();
        expect(foo).toBeDefined();
    });
});
```

Do you have any guidance on how to interpret `journey did not finish executing`?

Edit: removed unimportant `import` statement to reproduce the issue.

### PR DESCRIPTION
**Pull request by vigneshshanmugam**
### Summary 

There are lot of changes on this PR and needs to be done at the same time so we can get all issues addressed at with sourcemap generation and retrieving them with proper stack trace, Will cover one by one to make it easier for the reviewers

### Running test suites

+ Previously, the agent relied on `ts-node` for transpilation of the journey files which could be written in `.ts, .js or .mjs` files and prior to running the tests, the transformation of these files happens on the fly and resolved internally to the trasnspiled version. This was a good solution for the time being, but turned out to be problematic given the sourcemap locations were wrong most of the times as `ts-node` automatically installs the `source-map-support` module and we will not have granular access to the generated source files. 
+ Now we rely on the custom transformation using `esbuild` to achieve similar behaviour, which means we get the same benefits like automatic transpilation of the TS files, configs, etc. with also granular control of the sourcemap that got generated for those files. This becomes super handy as we can reuse those sourcemaps and maps to those files after bundling using esbuild. 

### Bundling
+ Previously, we used esbuild in Bundling mode with a custom `Synthetics` plugin that resolved all the journey and imported modules without any transpilation and bundled only the external `node_modules`. This was done for main two reasons
   - Reduce the bundle size of the generated journey file - No sourcemaps
   - Replicate the file structure so the errors are captured with proper stacktrace information. 

This method turned our to be really complex and we would need to handle some weird edge cases which is on another PR here - https://github.com/elastic/synthetics/pull/673

+ The current solutions relies on `Esbuild` to bundle the full journey in a single `bundled file` and use sourcemaps that are inlined to map it to the source when errors are thrown. One thing to note here is we set `sourceContents: false` as we dont need the full source content right now as we are not replicating the file structure on the Kibana side, plus removing it also reduces the size by 90%. The rest is just base64 content mappings. 

### Making sourcemaps work
+ To ensure correct sourcemap gets generated during the running suites phase and also post bundling phase, we have to have a single source of truth. Meaning
   - Single tool should be used for transpiling the source journey files and bundling.
   - `source-map-support` should be installed only once, otherwise mapping back to the original source position would become problematic. 
+ To avoid these problems, we use the custom transformation phase and store the sourcemaps in memory for each of the generated files and use them for retrieving later to find which journey was executed. 
+ We rely on the correct source map also for `journey, step, monitor` functions as we use them to identify all the duplicate monitors and report them as errors. 

#### Bug fixes 
+ fix #652 
+ Helps us move forward to get #596 done

Failed journey as discussed on the issue #652 

Screen Shot 2022-12-20 at 11 36 42 AM

### CODE REVIEW

File: src/push/plugin.ts [hunk 1]
```diff
@@ -23,145 +23,37 @@
...
-
-        if (
-          !isBare(args.path) ||
-          args.importer.includes('/node_modules/') ||
-          isLocalSyntheticsModule(args.importer)
-        ) {
-          return;
-        }
-
+        // Ignore entry points as these refer to the journey files
```
Comment (andrewvc) [inline src/push/plugin.ts:47]: love the simplification here

**Review by andrewvc — CHANGES_REQUESTED** [review]
This all looks great, but there's complex implementation details here that could really use comments. The why and how are well fleshed out in the issue description, can you copy that content into the codebase as comments as well?

File: __tests__/push/bundler.test.ts [hunk 2]
```diff
@@ -33,6 +33,34 @@ import { Bundler } from '../../src/push/bundler';
...
 const journeyFile = join(PROJECT_DIR, 'bundle.journey.ts');
 
+const setup = async () => {
+  await writeFile(
+    journeyFile,
+    `import {journey, step, monitor} from '@elastic/synthetics';
+import isPositive from 'is-positive';
+import utils from "./utils"
+journey('journey 1', () => {
+  // avoid dead code elimination
```
[the remark this answers is not in the document]
Comment (vigneshshanmugam) [reply __tests__/push/bundler.test.ts:43]: If we dont call utils. Then the code gets eliminated as its unused, which is why I added that comment. Seems like that did not help. I will add more detailed comment in this case.

File: src/core/transform.ts [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+};
+
+const getLoader = (filename: string) => {
+  const ext = path.extname(filename);
+  return LOADERS[ext] || 'default';
+};
+
+export function commonOptions(): CommonOptions {
+  return {
+    minify: false,
```
[the remark this answers is not in the document]
Comment (vigneshshanmugam) [reply src/core/transform.ts:67]: Added comments.

**Review by andrewvc — APPROVED** [review]
LGTM

**Pull request merged**